### ISSUE TITLE
Failed push/pull leads to missing configs and a broken dev environment

### ISSUE THREAD

**Issue by PeteDuncanson** [aws-amplify/amplify-cli#6047]
**Describe the bug**

I think my local setup has got into an unknown state. It regularly complains about missing files and missing environments. I have to cut and paste them from a team mate to get it to "work" again but it can quickly and randomly fall back into a unknown state.

I'm not sure if this is related to my issues with VSCode #6028? Could it be that when I push and get a error half way through Amplify has already deleted some files or folders and then failed in an unknown state halfway through? Then on repeat runs it can't find config that it needs?

Running `amplify pull` tells me my environment doesn't exist and recommends I try to "init" it again. Running `amplify init` then fails due to a missing config (which I've never deleted), oddly it does know about the missing environment though so something not right there, seems push and inti aren't talking to the same point of truth?

**Amplify CLI Version**

v4.32

**To Reproduce**

This seems to have started since the CLI got half way through a push and then failed with a permissions error (see #6028). That then means that one or several config files are missing (cloudformation-template.json, backend-config.json and amplify-meta.json). This gets me in a stuck state, can't recover from this through the CLI.

**Expected behavior**

Ideally it doesn't get into this state in the first place. Or if it can't find the configs then offers to go get them from the cloud? It should be able to recover from this state. This isn't caused by me deleting things (if I did that I deserve what I get) it seems to be caused by Amplify doing its thing (a push or pull) and failing half way leaving me with a project in an unknown state. All the information needed to refresh this state is I think up in the cloud. 

**Screenshots**

So shown here is me trying to do a push, this fails as I've touched to keys in my schema at the same time but it rolls back, you can see the error for that right at the top of the first screen shot. Then I try again after editing my schema to remove one of the key changes, this fails with an odd error about some auth settings being missing, I've not touched anything to do with those. I then try to do a pull to get back in a known state and thats when it says "what environment?" so I try init as hinted at in the error and that blows up too. Whole solution now is broken unless I cut and paste over the config from a team mate.

![image](https://user-images.githubusercontent.com/1674590/101005239-b2e9e800-3558-11eb-9f2f-0a42e24e3eb2.png)
![image](https://user-images.githubusercontent.com/1674590/101005535-c8f7a880-3558-11eb-8dc4-5daedd2d477c.png)
![image](https://user-images.githubusercontent.com/1674590/101004898-9a79cd80-3558-11eb-8ae6-5cffac6c21b4.png)

**Desktop (please complete the following information):**
 - OS: Windows
 - Node Version. v12.18.4

**Additional context**

**Comment by PeteDuncanson** [comment]
To add to this it seems it has deleted all my cloud templates etc?

![image](https://user-images.githubusercontent.com/1674590/101009222-da8d8000-3559-11eb-8c18-48c5d32aff09.png)

**Comment by PeteDuncanson** [comment]
I've taken the step of recreating my dev environment into a fresh folder, fresh check out from github and then ran `amplify pull --restore`

That seems to have sorted out my issues with missing files etc. and I can now push. Hope that helps anyone who might have the same issues.

There is still  the issue though of how can it get into that state in the first place? This issue could be closed as the above will sort of get you out of it, but might be worth a bit of a think if there is a way to prevent it getting this broken in the first way or is there a way it can recover from it, even if its just a warning message pretty much telling them to do the above worst case?

**Comment by SwaySway** [comment]
@PeteDuncanson Based on the screenshot, are the only files being deleted within api? If so are you running mock before deploying?

**Comment by PeteDuncanson** [comment]
@SwaySway no I don't have mock running. They aren't all within API, two are in /amplify/backend

**Comment by SwaySway** [comment]
It's possible this is a windows specific issue. Does your teammate use windows as well? If so have they also encountered this issue?

**Comment by PeteDuncanson** [comment]
He's not had the same issue no. I'm pretty sure I messed up my setup somehow with a half way pull or push or something which has got it into an unknown state, I think that is more likely that just being a windows thing?

**Comment by SwaySway** [comment]
I think this issue is most likely related to the upload not happening due the files not being able to be read. Do you have read share enabled for files? Win Docs [CreateFileA](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea)

This doesn't seem like a separate issue from #6028. Closing this issue in favor of that one.

**Comment by PeteDuncanson** [comment]
So this occurred again today, had a file open (a Lambda this time) in VSCode and it errors with the attached. At that point it has deleted 175 files that are in source and a few config files that aren't. My setup is now broken again and only fixable by  

![image](https://user-images.githubusercontent.com/1674590/101153748-bc3c8880-361c-11eb-94b4-a185e3cc3c4c.png)
snipped out all the pull output
![image](https://user-images.githubusercontent.com/1674590/101153806-c8c0e100-361c-11eb-9972-79f290c743f2.png)

Without that amplify-meta.json file my dev setup is dead, can't get out of this situation without cut and pasting in a copy of the file from my team mate. Pondering adding this file to source control at this rate.

**Comment by PeteDuncanson** [comment]
@SwaySway additionally I don't know what you mean about having "read share enabled" or that file you linked to? Got any more context?

I'm just stumped as to why it would be deleting my whole environment and then erroring about access issues when VSCode is open. To prevent it could we not check for all the access permissions first before doing any deleting? Or move the files to a temp folder so they can be copied back if anything goes wrong? Its the fact it keeps getting into an unknown state that worries me.

I've disabled all non-essential extensions in VSCode just now too in the hope it might be one of those that is keeping a handle to a file or something. Will keep an eye on it today and see how it goes.

Finally I've added in amplify-meta.json into source for now and see how it goes, there are only two of us on the team, if it causes issues I'll revert.
